server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -386,6 +386,18 @@ public short groupMetadataValueVersion() {
         }
     }
 
+    public short offsetCommitValueVersion() {
+        if (isLessThan(MetadataVersion.IBP_2_1_IV0)) {
+            return 1;
+        } else if (isLessThan(MetadataVersion.IBP_2_1_IV1)) {
+            return 2;
+        } else {
+            // Serialize with the highest supported non-flexible version
+            // until a tagged field is introduced or the version is bumped.
+            return  3;
+        }
```

Comment:
This is taken from [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L1088).

Reply:
>  // Serialize with the highest supported non-flexible version
            // until a tagged field is introduced or the version is bumped.
            
This comment confused me a bit. Do we plan to manually update this method when new versions come in? Why is there a callout for flexible versions?

Reply:
Also do we expect to have lower IBPs if this is used for the new group coordinator? I didn't think we could go that low for kraft.

Reply:
> > // Serialize with the highest supported non-flexible version
> > // until a tagged field is introduced or the version is bumped.
> 
> This comment confused me a bit. Do we plan to manually update this method when new versions come in? Why is there a callout for flexible versions?

This is related to https://cwiki.apache.org/confluence/display/KAFKA/KIP-915%3A+Txn+and+Group+Coordinator+Downgrade+Foundation. Basically, we have already introduced the version 4 of this record to support tagged fields that may be added in the future. However, we don't want to use version 4 yet.

When a new version is released or tagged fields are added, we will change the logic here to use the correct version based on the metadata version.

Reply:
> Also do we expect to have lower IBPs if this is used for the new group coordinator? I didn't think we could go that low for kraft.

You're right. We should never hit those old versions as the new group coordinator only runs in kraft mode. I think that we will be able to clean up those when we will remove all the old ZK related code and version in the future. However, at the moment, we went conservative and kept the logic in order to be consistent in the current code base. Do you agree with this strategy?

Reply:
Ok. Sounds fair. Do we have a jira to do such a cleanup? Just don't want it to be forgotten.

As for version 4, is that coming up in a future PR?

Reply:
I suppose that those cleanups will happen when we remove ZK from the code base. I am not aware of any jiras for this at the moment.

Yes. Version 4 will come later this year. We are waiting a bit in order to have a downgrade window of a couple of releases.